### PR TITLE
[DiskBBQ] Make KMeans related classes generic in anticipation of byte[] support

### PR DESCRIPTION
**Pull request by john-wagster**
Related to Phase 2 of the broader effort (https://github.com/elastic/elasticsearch/issues/146407) to add 'byte' element type to diskbbq. 

This adds support for extracting float specific functionality into a FloatOps class and then adding generic parameters to the Kmeans classes in anticipation of supporting ByteOps subsequently.  This is done to try to minimize the impact of introducing byte[] to the Kmeans classes where possible.

This particular PR is in conjunction with (https://github.com/elastic/elasticsearch/pull/149771) and the second step in extending the Kmeans algorithms to generically handle byte[].  It's a follow up to https://github.com/elastic/elasticsearch/issues/148436 where we extended OSQ related functions to support byte[].

There is a branch that has all of these changes and preps subsequent changes to add ByteOps support which I've linked here if it helps for context: https://github.com/elastic/elasticsearch/compare/main...john-wagster:elasticsearch:generic-kmeans-clustering?expand=1.  This will make up the next PR in the set to complete Phase 2.

There is an even larger scoped branch with the full PoC of 'byte' element type support here: https://github.com/elastic/elasticsearch/compare/main...john-wagster:byte_for_diskbbq3?expand=1
Preliminary results on this PoC show consistent recall numbers in line with naively casting bytes from the initial PoC

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java [hunk 1]
```diff
@@ -189,8 +180,8 @@ protected void innerCluster(ClusteringFloatVectorValues vectors, KMeansIntermedi
 
         float[] cumulativeClusterWeights = new float[k]; // maintains soft cluster counts for each cluster.
                                                          // Used to compute the learning rate in the SGD update of the centroids
-        for (int idx = 0; idx < k; idx++) {
-            if (assignments[idx] == -1) {
+        for (int idx = 0; idx < n; idx++) {
+            if (assignments[idx] != -1) {
```
Comment (john-wagster) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java:184]: this is one of the few changes that is not a refactor; I believe it to be a bug in that previously we would try to count missing assignments and should have been counting actual prior assignments

Comment (marianotepper) [reply server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java:184]: Nice catch!

**Comment by marianotepper** [comment]
As I go through the review I have a general comment: I'm surprised to see that centroids are V[]. I would expect input vectors to be generic, but centroids should remain float[]. Why would we change the centroids too?

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java [hunk 2]
```diff
@@ -109,59 +121,38 @@ private static int argMin(float[] vector) {
...
+        CentroidOps.FloatOps floatOps = (CentroidOps.FloatOps) ops;
         for (int idx = 0; idx < vectors.size(); idx++) {
-            float[] vec = vectors.vectorValue(idx);
+            V vec = vectors.vectorValue(idx);
             int k = assignments[idx];
             cumulativeClusterWeights[k]++;
-            float learning_rate = 1.f / (cumulativeClusterWeights[k] + learningRateShift);
-            ESVectorUtil.linearCombination(learning_rate, vec, 1 - learning_rate, centroids[k]);
+            float learningRate = 1.f / (cumulativeClusterWeights[k] + learningRateShift);
+            floatOps.linearCombination(learningRate, (float[]) vec, 1 - learningRate, (float[]) centroids[k]);
```
Comment (marianotepper) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java:138]: Are these casts to float[] safe?

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    /**
+     * Compute {@code diffs[d] = vector[d] - centroid[d]} as floats (for SOAR residuals).
+     * Always produces {@code float[]} regardless of vector type, because the SOAR formula
+     * operates in float space.
+     */
+    void computeDiffs(V vector, V centroid, float[] diffs);
+
+    /**
+     * Convert centroids to {@code float[][]} for use with float-only subsystems (e.g. {@link NeighborHood}).
+     * For {@link FloatOps} this is a no-op cast.
```
Comment (marianotepper) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java:98]: I think it would be helpful to add a clarifying comment for the general non-FloatOps case and its memory/compute implications.

**Comment by john-wagster** [comment]
> As I go through the review I have a general comment: I'm surprised to see that centroids are V[]. I would expect input vectors to be generic, but centroids should remain float[]. Why would we change the centroids too?

@marianotepper this was a conversation with @benwtrent earlier on and the thinking here (feel free to correct me Ben) is that we really want to show the value of indexing docs with lower heap which we only get from converting as many of the float[] as possible including the centroids.

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+import org.elasticsearch.simdvec.MathUtils;
+
+/**
+ * Encapsulates all vector/centroid-type-specific arithmetic for k-means clustering.
+ * <p>
+ * Currently provides {@link FloatOps} for {@code float[]} vectors/centroids.
+ *
+ * @param <V> the array type for vectors and centroids ({@code float[]})
+ */
+public sealed interface CentroidOps<V> permits CentroidOps.FloatOps {
```
Comment (marianotepper) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java:22]: Is it worth having unit tests for this class now that it has been refactored?

**Review by marianotepper — COMMENTED** [review]
Nice refactoring. Very clean, much easier to follow. Thanks!

**Comment by marianotepper** [comment]
> > As I go through the review I have a general comment: I'm surprised to see that centroids are V[]. I would expect input vectors to be generic, but centroids should remain float[]. Why would we change the centroids too?
> 
> @marianotepper this was a conversation with @benwtrent earlier on and the thinking here (feel free to correct me Ben) is that we really want to show the value of indexing docs with lower heap which we only get from converting as many of the float[] as possible including the centroids.

Did you evaluate the clustering with byte[] centroids? I would be surprised if this does not affect the stochastic gradient descent algorithm, as we are moving the centroids by relatively small amounts, which may be lost for byte[].

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java (hunk 2 above)
Comment (john-wagster) [reply server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java:138]: yes in that we will know by this point that we are dealing with a float vector.  A subsequent PR here will make this a little more clear but I didn't want to convolute this one where we will check for the type of CentroidOps and potentially handle different logic accordingly.  I can temporarily gate it if you have concerns. 

Here's what this will look like after that iteration where floatCentroidsShadow is a set of temporary float[] and is only defined for the 'byte' path: 
```
        if (floatCentroidsShadow != null) {
            // Byte path: SGD in float precision, round to byte for distance computation
            int dim = vectors.dimension();
            for (int idx = 0; idx < vectors.size(); idx++) {
                byte[] vec = (byte[]) vectors.vectorValue(idx);
                int k = assignments[idx];
                cumulativeClusterWeights[k]++;
                float lr = 1.f / (cumulativeClusterWeights[k] + learningRateShift);
                ESVectorUtil.linearCombination(lr, vec, 1 - lr, floatCentroidsShadow[k]);
                // Round back to byte centroid for distance computation
                byte[] byteCentroid = (byte[]) centroids[k];
                for (int d = 0; d < dim; d++) {
                    byteCentroid[d] = (byte) Math.clamp(Math.round(floatCentroidsShadow[k][d]), -128, 127);
                }
            }
        } else {
            // Float path
            CentroidOps.FloatOps floatOps = (CentroidOps.FloatOps) ops;
            for (int idx = 0; idx < vectors.size(); idx++) {
                V vec = vectors.vectorValue(idx);
                int k = assignments[idx];
                cumulativeClusterWeights[k]++;
                float learningRate = 1.f / (cumulativeClusterWeights[k] + learningRateShift);
                floatOps.linearCombination(learningRate, (float[]) vec, 1 - learningRate, (float[]) centroids[k]);
            }
        }
```

Thoughts and further concerns here are more than welcome

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java (hunk 3 above)
Comment (john-wagster) [reply server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java:98]: this is the comment I have in the subsequent PR that will include ByteOps

```
    /**
     * Convert centroids to {@code float[][]} for use with float-only subsystems (e.g. {@link NeighborHood}).
     * For {@link FloatOps} this is a no-op cast. For {@link ByteOps} this widens each byte to float.
     */
```

Do you think that's sufficient or we should have more details here.  And are you ok if that goes in on the subsequent PR?  I'm ok either way but obviously there's no ByteOps yet so didn't want to confuse anything here for folks.

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java (hunk 4 above)
Comment (john-wagster) [reply server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java:22]: My gut reaction is similar to some of the other comments so far; I think when we get to adding ByteOps in the next PR this may make more sense particularly to check for things like overflow or rounding errors.  Most of these functions will be (for byte[]) and are (for float[]) which already are tested by *KMeans*Tests or KMeansResultTests already and those that delegate to ESVectorUtil functions already have tests too.  Please feel free to push back though.

**Comment by benwtrent** [comment]
@marianotepper we don't have a `byte -> float` SIMD kernel, and keeping the vectors as `byte[]` is pretty important for throughput costs. 

The clusters as they are formed are indeed `float[]` and then should be rounded to `byte[]` after their transformation. I wonder if the minibatch changes can work on `float[]` centroids at a time, but when the centroid pass is finalized for an iteration, they are transformed to `byte[]`.

**Comment by marianotepper** [comment]
Good points @benwtrent.

Upon reflection, we are also quantizing the clusters once the index is built to a lower precision than byte (but with a slightly better quantizer). Thanks @john-wagster for confirming this offline. Taking this into consideration, I think this is fine. Interestingly, this also explains why we can get away with so few iterations/epochs over the data when running the clsutering.

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java (hunk 3 above)
Comment (marianotepper) [reply server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/CentroidOps.java:98]: Yeap

**Review by marianotepper — APPROVED** [review]
Thanks for the answers @john-wagster. Nice work!

File: server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java [hunk 5]
```diff
@@ -27,16 +27,19 @@
  * assigning each vector.
  * The regularization augments each distance by a component proportional to the current size of the corersponding cluster,
  * so we scale this additional term by the median distance to centroids in the mini batch to balance both terms.
+ *
+ * @param <V> the array type for vectors and centroids ({@code float[]} or {@code byte[]})
  */
-abstract class BalancedASKMeansLocal extends KMeansLocal {
+abstract class BalancedASKMeansLocal<V> extends KMeansLocal<V> {
```
Comment (tteofili) [inline server/src/main/java/org/elasticsearch/index/codec/vectors/cluster/BalancedASKMeansLocal.java:33]: ah, generics! ❤️

**Review by tteofili — APPROVED** [review]
LGTM
